modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/incoming/IncomingSnapshotCopier.java

```diff
@@ -113,8 +147,174 @@ public SnapshotReader getReader() {
+            future.completeExceptionally(copierException(e, RaftError.EIO, "Error recreating partition"));
+        } catch (Throwable t) {
+            future.completeExceptionally(t);
+        } finally {
+            cancelBusyLock.leaveBusy();
+        }
+    }
+
+    private void requestSnapshotMetaAsync(ClusterNode snapshotSenderNode) {
+        if (!cancelBusyLock.enterBusy()) {
+            return;
+        }
+
+        try {
+            partitionSnapshotStorage.outgoingSnapshotsManager().messagingService().invoke(
+                    snapshotSenderNode,
+                    MSG_FACTORY.snapshotMetaRequest().id(snapshotUri.snapshotId).build(),
+                    NETWORK_TIMEOUT
+            ).whenComplete((networkMessage, throwable) -> {
+                if (throwable != null) {
+                    future.completeExceptionally(copierException(throwable, RaftError.EIO, "Failed to request snapshot meta"));
+                } else {
+                    snapshotMeta = ((SnapshotMetaResponse) networkMessage).meta();
+
+                    requestSnapshotMvDataAsync(snapshotSenderNode);
+                }
+            });
+        } catch (Throwable t) {
+            future.completeExceptionally(t);
+        } finally {
+            cancelBusyLock.leaveBusy();
+        }
+    }
+
+    private void requestSnapshotMvDataAsync(ClusterNode snapshotSenderNode) {
+        if (!cancelBusyLock.enterBusy()) {
+            return;
+        }
+
+        try {
+            partitionSnapshotStorage.outgoingSnapshotsManager().messagingService().invoke(
+                    snapshotSenderNode,
+                    MSG_FACTORY.snapshotMvDataRequest().id(snapshotUri.snapshotId).build(),
+                    NETWORK_TIMEOUT
+            ).whenComplete((networkMessage, throwable) -> {
+                if (throwable != null) {
+                    future.completeExceptionally(copierException(throwable, RaftError.EIO, "Failed to load mv partition data"));
+                } else {
+                    SnapshotMvDataResponse snapshotMvDataResponse = ((SnapshotMvDataResponse) networkMessage);
+
+                    handleVersionChains(snapshotMvDataResponse.rows());
+
+                    if (!snapshotMvDataResponse.finish()) {
+                        requestSnapshotMvDataAsync(snapshotSenderNode);
+                    } else {
+                        completeCopySnapshot();
+                    }
+                }
+            });
+        } catch (Throwable t) {
+            future.completeExceptionally(t);
+        } finally {
+            cancelBusyLock.leaveBusy();
+        }
+    }
+
+    private void completeCopySnapshot() {
+        if (!cancelBusyLock.enterBusy()) {
+            return;
+        }
+
+        try {
+            SnapshotMeta snapshotMeta0 = snapshotMeta;
+
+            assert snapshotMeta0 != null;
+
+            partitionSnapshotStorage.partition().lastAppliedIndex(snapshotMeta0.lastIncludedIndex());
```

Comment:
I don't think I see a part when you set applied index to -1. Why did you decided to ignore it?

Reply:
Tried to fix it.